File: strategy/generator/signal/trend_candle.py
from dataclasses import dataclass

from core.models.candle import TrendCandleType
from core.models.parameter import CategoricalParameter, Parameter

from .base import Signal, SignalType


@dataclass(frozen=True)
class TrendCandleSignal(Signal):
    type: SignalType = SignalType.TrendCandle
    candle: Parameter = CategoricalParameter(TrendCandleType)

File: ta_lib/strategies/signal/src/trend_candle.rs
use base::prelude::*;
use core::prelude::*;
use shared::{trend_candle_indicator, TrendCandleType};

const DEFAULT_LOOKBACK: usize = 13;

pub struct TrendCandleSignal {
    candle: TrendCandleType,
}

impl TrendCandleSignal {
    pub fn new(candle: TrendCandleType) -> Self {
        Self { candle }
    }
}

impl Signal for TrendCandleSignal {
    fn lookback(&self) -> usize {
        DEFAULT_LOOKBACK
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        trend_candle_indicator(&self.candle, data)
    }
}
